### PR TITLE
system_instruction_processor updates

### PR DESCRIPTION
**Pull request by rob-solana**
#### Problem
 create_account doesn't check if the 'to' account is empty before
 assigning the owner, which would allow an attacker to steal any balance by
 assigning to their own program

 #### Summary of Changes
 * verify that `to` is empty before assigning, this is a band-aid, need to sign `to` #6449 
 * add support for zero transfers
 * increased test coverage
 * removed redundant check for source program id (is already enforced by runtime)
 * whacked create_system_account(), which tests were using like transfer(), expecting that to work
      multiple times (which this band-aid now prevents)

Fixes #

### CODE REVIEW

**Comment by rob-solana** [comment]
I started on this so that I could do stake.split() with:
* a single signature
* no extra lamports for account creation

I ended up with some more stuff.

**Comment by mvines** [comment]
> create_account doesn't check if the 'to' account is empty before
assigning the owner, which would allow an attacker to steal any balance by
assigning to their own program

Doesn't the existing `system_program::check_id(&to.account.owner)` check prevent this?
https://github.com/solana-labs/solana/blob/35ed432d1a4d360a07c72a423bc80a3b84226072/runtime/src/system_instruction_processor.rs#L30

I can't steal the funds, I can only sneak in another create_account transaction ahead of yours as a DoS

**Comment by rob-solana** [comment]
> > create_account doesn't check if the 'to' account is empty before
> > assigning the owner, which would allow an attacker to steal any balance by
> > assigning to their own program
> 
> Doesn't the existing `system_program::check_id(&to.account.owner)` check prevent this?
> https://github.com/solana-labs/solana/blob/35ed432d1a4d360a07c72a423bc80a3b84226072/runtime/src/system_instruction_processor.rs#L30
> 
> I can't steal the funds, I can only sneak in another create_account transaction ahead of yours as a DoS

no.  consider the case of a transfer of lamports to a system account.  at any time after that transfer anyone can assign the owner (e.g. to a "withdraw" program)

File: runtime/src/storage_utils.rs [hunk 1]
```diff
@@ -110,26 +110,20 @@ pub(crate) mod tests {
         let bank = Arc::new(bank);
         let bank_client = BankClient::new_shared(&bank);
 
-        bank_client
-            .transfer(10, &mint_keypair, &replicator_pubkey)
-            .unwrap();
         let message = Message::new(storage_instruction::create_storage_account(
```
Comment (rob-solana) [inline runtime/src/storage_utils.rs]: @sagar-solana this is why you're invited

**Comment by garious** [comment]
This PR, more than most all others, needs a GitHub issue, to be assigned to a milestone, and to be referenced in the PR description.

**Comment by rob-solana** [comment]
CC #6449

**Pull request merged**